An example for a quick start guide

yoni206 opened the pull request:
Co-authored-by: Aina Niemetz <aina.niemetz@gmail.com>

aniemetz commented on examples/api/cpp/quickstart.cpp:
```diff
@@ -0,0 +1,155 @@
+  solver.setOption("produce-unsat-cores", "true");
+
+  // The simplest way to set a logic for the solver is to choose "ALL".
+  // This enables all logics in the solver.
```
```suggestion
  // This enables all logics in the solver.
  // Alternatively, "QF_ALL" enables all logics without quantifiers.
```

aniemetz commented on examples/api/cpp/quickstart.cpp:
```diff
@@ -0,0 +1,155 @@
+
+  // In this example, we will define constraints
+  // over reals and integers.
+  // Hence we first obtain the corresponding sorts.
```
```suggestion
  // Hence, we first obtain the corresponding sorts.
```

aniemetz commented on examples/api/cpp/quickstart.cpp:
```diff
@@ -0,0 +1,155 @@
+  Sort realSort = solver.getRealSort();
+  Sort intSort = solver.getIntegerSort();
+
+  // x,y will be real variables, while
```
```suggestion
  // x and y will be real variables, while
```

aniemetz commented on examples/api/cpp/quickstart.cpp:
```diff
@@ -0,0 +1,155 @@
+  Sort intSort = solver.getIntegerSort();
+
+  // x,y will be real variables, while
+  // a,b will be integer variables.
```
```suggestion
  // a and b will be integer variables.
```

aniemetz commented on examples/api/cpp/quickstart.cpp:
```diff
@@ -0,0 +1,155 @@
+  Term a = solver.mkConst(intSort, "a");
+  Term b = solver.mkConst(intSort, "b");
+
+  // Our constraints regarding x,y will be:
```
```suggestion
  // Our constraints regarding x and y will be:
```

aniemetz commented on examples/api/cpp/quickstart.cpp:
```diff
@@ -0,0 +1,155 @@
+  Term b = solver.mkConst(intSort, "b");
+
+  // Our constraints regarding x,y will be:
+  // '''
```
```suggestion
  //
```

aniemetz commented on examples/api/cpp/quickstart.cpp:
```diff
@@ -0,0 +1,155 @@
+  //   (2)  0 < y
+  //   (3)  x + y < 1
+  //   (4)  x <= y
+  // '''
```
```suggestion
  //
```

aniemetz commented on examples/api/cpp/quickstart.cpp:
```diff
@@ -0,0 +1,155 @@
+  Term zero = solver.mkReal(0);
+  Term one = solver.mkReal(1);
+
+  // Next, we construct the term x+y
```
```suggestion
  // Next, we construct the term x + y
```

aniemetz commented on examples/api/cpp/quickstart.cpp:
```diff
@@ -0,0 +1,155 @@
+  // Next, we construct the term x+y
+  Term xPlusY = solver.mkTerm(PLUS, x, y);
+
+  // Next, we continue with the compound terms, that
```
This looks like a part of the sentence is missing?

aniemetz commented on examples/api/cpp/quickstart.cpp:
```diff
@@ -0,0 +1,155 @@
+  // Next, we continue with the compound terms, that
+
+  // Now we can define the constraints.
+  // They employ +, <=, and <.
```
```suggestion
  // They use the operators +, <=, and <.
```

aniemetz commented on examples/api/cpp/quickstart.cpp:
```diff
@@ -0,0 +1,155 @@
+  // They employ +, <=, and <.
+  // In the API, these are denoted by
+  // PLUS, LEQ, and LT.
+  // A list of operators is available in:
```
```suggestion
  // A list of available operators is available in:
```

aniemetz commented on examples/api/cpp/quickstart.cpp:
```diff
@@ -0,0 +1,155 @@
+  solver.assertFormula(constraint4);
+
+  // Check if the formula is satisfiable, that is,
+  // are there real x,y,z that satisfy all the
```
```suggestion
  // are there real values for x,y,z that satisfy all the
```

aniemetz commented on examples/api/cpp/quickstart.cpp:
```diff
@@ -0,0 +1,155 @@
+
+  // The result is either SAT, UNSAT, or UNKNOWN.
+  // In this case, it is SAT.
+  std::cout << "result:" << r1 << std::endl;
```
```suggestion
  std::cout << "expected: sat" << std::endl;
  std::cout << "result:" << r1 << std::endl;
```

aniemetz commented on examples/api/cpp/quickstart.cpp:
```diff
@@ -0,0 +1,155 @@
+  // In this case, it is SAT.
+  std::cout << "result:" << r1 << std::endl;
+
+  // We can get actual values for x,y that
```
```suggestion
  // We can get the values for x and y that
```

aniemetz commented on examples/api/cpp/quickstart.cpp:
```diff
@@ -0,0 +1,155 @@
+  std::string xStr = xVal.getRealValue();
+  std::string yStr = yVal.getRealValue();
+
+  // now we can convert the values to cpp types
+  // using standard cpp conversion functions.
+  double xDouble = std::stod(xStr);
+  double yDouble = std::stod(yStr);
+
+  std::cout << "solution for x: " << xDouble << std::endl;
+  std::cout << "solution for y: " << yDouble << std::endl;
```
You could just print `xStr` and `yStr` without having to convert them to double, no?

aniemetz commented on examples/api/cpp/quickstart.cpp:
```diff
@@ -0,0 +1,155 @@
+  std::cout << "solution for y: " << yDouble << std::endl;
+
+  // Next, we will check satisfiability of the same formula,
+  // only this time over integer variables a,b
```
```suggestion
  // only this time over integer variables a and b.
```

aniemetz commented on examples/api/cpp/quickstart.cpp:
```diff
@@ -0,0 +1,155 @@
+  // Next, we will check satisfiability of the same formula,
+  // only this time over integer variables a,b
+
+  // We start by reseting the solver to flush
```
```suggestion
  // We start by resetting assertions added to the solver.
```

aniemetz commented on examples/api/cpp/quickstart.cpp:
```diff
@@ -0,0 +1,155 @@
+  // only this time over integer variables a,b
+
+  // We start by reseting the solver to flush
+  // previous assertions.
```
```suggestion
```

aniemetz commented on examples/api/cpp/quickstart.cpp:
```diff
@@ -0,0 +1,155 @@
+  solver.resetAssertions();
+
+  // Next, we assert the same assertions above with integers,
+  // in an abbreviated manner.
```
What is the abbreviated manner?

aniemetz commented on examples/api/cpp/quickstart.cpp:
```diff
@@ -0,0 +1,155 @@
+  Result r2 = solver.checkSat();
+
+  // This time the formula is unsatisfiable
+  std::cout << "result: " << r2 << std::endl;
```
```suggestion
  std::cout << "expected: unsat" << std::endl;
  std::cout << "result: " << r2 << std::endl;
```

aniemetz reviewed (commented):
Overall, the example looks really good.
What we might want to add:

* get value of an expression that is not part of an assertion, e.g., x * y
* create an indexed term, maybe? maybe this is out of scope for this one, though

Can you also maybe rewrap the comments a bit to better make use of the 80 char length? Some are broken after half of the line.

yoni206 commented:
Thanks! I added a `get-value` for `x-y` (which is not in the original formula).

Regarding indexed operator, personally I wouldn't put them in the very first example, but I don't feel strongly about it so let me know if you prefer I add these.

Regarding the conversion to `double` before printing, I now added some lines that manipulate the `double`s (my goal was to show how to get actuall arithmetic values and not strings). 

All other comments are addressed.

aniemetz commented on examples/api/cpp/quickstart.cpp:
```diff
@@ -0,0 +1,167 @@
+  // However, for more complex terms,
+  // it is easier to let the solver do the evaluation.
+  double xMinusYComputed = xDouble - yDouble;
+  if (xMinusYComputed == xMinusYDouble) {
+    std::cout << "computed correctly" << std::endl;
+  } else {
+    std::cout << "computed incorrectly" << std::endl;
+  }
```
Format.

aniemetz commented on examples/api/cpp/quickstart.cpp:
```diff
@@ -0,0 +1,167 @@
+  double xMinusYDouble = std::stod(xMinusYStr);
+
+  // Another way to independently compute the value of x and y would be using
+  // ordinary cpp minus operator, rather than asking the solver.
```
```suggestion
  // the ordinary cpp minus operator instead of asking the solver.
```

aniemetz reviewed (commented):
Just two minor things, else LGTM.
